scheduler/src/cook/authorization.clj

```diff
@@ -0,0 +1,180 @@
+  `:authorization-fn` config file key to it. 
+
+  For example, you could consult an external authorization system that
+  manages user credits or billing, or you could implement a \"teams\"
+  feature, where users are allowed to modify their teammates' jobs.
+
+"
+  (:require [cook.global-state :refer [global-state]]
+            [cook.util :refer [lazy-load-var]]
+            [clojure.tools.logging :as log]
+            [plumbing.core :refer (fnk defnk)]))
+
+
+;; The defonce is a kludge to work around
+;; https://groups.google.com/forum/#!topic/clojure/SYYYwZIrFiY -
+;; reloading a protocol causes existing objects of that type to become
+;; invalid as instances of the new protocol.
+;;
+;; If you redefine this protocol at runtime, you'll have to manually
+;; remove the defonce. Other than that, this should not cause issues.
+(defonce __ownable
+  (defprotocol Ownable
+    "Functions to be defined on types that are able to be owned by a specific user."
+    (owner [this] "Returns the username that owns this object.")))
+
+(def system (reify Ownable (owner [this] ::system)))
+
+;;
+;; Authorization functions
+;; 
+
+
+(defn open-auth
+  "This authorization function allows any user to do anything to any
+  object. It is intended for testing and development."
+  [settings user verb object]
+  true)
+
+
+(defn configfile-admins-auth
+  "This authorization function consults the set of usernames specified
+  in the :admins key of the config file.
+
+  Usernames in this set are administrators, who are allowed to do anything to any object.
+  Non-admins are only allowed to manipulate objects that they own."
+  [settings
+   ^String user
+   ^clojure.lang.Keyword verb
+   ^cook.authorization.Ownable object ]
+  (let [admins    (:admins settings)
+        is-admin? (contains? admins user)
+        owner     (owner object)]
+    (log/debug "[configfile-admins-auth] Checking whether user" user
+               "may perform" verb
+               "on object" (str object) "."
+               "Admins are:" admins)
+    (cond is-admin? (do
+                      (log/debug "[configfile-admins-auth] User" user "is an admin, allowing.")
+                      true)
+          (= owner user) (do
+                           (log/debug "[configfile-admins-auth] Object is owned by user, allowing.")
+                           true)
+          :else (do 
+                  (log/info "[configfile-admins-auth] Unauthorized access attempt: user" user
+                            "is not allowed to perform" verb "on" (str object ",") "denying.")
+                  false))))
+
+
+(defn is-authorized?
+  "Determines whether the given user can perform the given operation
+  on the given object, according to the authorization function
+  specified in the config file.
+
+  Returns true if allowed, else false."
+  ([^String user
+     ^clojure.lang.Keyword verb
+     ^cook.authorization.Ownable object] 
+     (is-authorized? (-> @global-state :settings) user verb object))
```

Comment:
I prefer to create a partial function in which we explicitly pass in the config instead of using global state. Is there a compelling reason to do it this way?

An alternative to creating a partial function in components would be to have a `setup-is-authorized?` which would take the settings and return a `is-authorized?` function with is properly setup.